worlds/witness/entity_hunt.py

```diff
@@ -0,0 +1,254 @@
+from collections import defaultdict
+from logging import debug, warning
+from pprint import pformat
+from typing import TYPE_CHECKING, Dict, List, Set, Tuple
+
+from .data import static_logic as static_witness_logic
+
+if TYPE_CHECKING:
+    from . import WitnessPlayerLogic, WitnessWorld
+
+
+DISALLOWED_ENTITIES_FOR_PANEL_HUNT = {
+    "0x03629",  # Tutorial Gate Open, which is the panel that is locked by panel hunt
+    "0x03505",  # Tutorial Gate Close (same thing)
+    "0x3352F",  # Gate EP (same thing)
+    "0x00CDB",  # Challenge Reallocating
+    "0x0051F",  # Challenge Reallocating
+    "0x00524",  # Challenge Reallocating
+    "0x00CD4",  # Challenge Reallocating
+    "0x00CB9",  # Challenge May Be Unsolvable
+    "0x00CA1",  # Challenge May Be Unsolvable
+    "0x00C80",  # Challenge May Be Unsolvable
+    "0x00C68",  # Challenge May Be Unsolvable
+    "0x00C59",  # Challenge May Be Unsolvable
+    "0x00C22",  # Challenge May Be Unsolvable
+    "0x0A3A8",  # Reset PP
+    "0x0A3B9",  # Reset PP
+    "0x0A3BB",  # Reset PP
+    "0x0A3AD",  # Reset PP
+}
+
+ALL_HUNTABLE_PANELS = [
+    entity_hex
+    for entity_hex, entity_obj in static_witness_logic.ENTITIES_BY_HEX.items()
+    if entity_obj["entityType"] == "Panel" and entity_hex not in DISALLOWED_ENTITIES_FOR_PANEL_HUNT
+]
+
+
+class EntityHuntPicker:
+    def __init__(self, player_logic: "WitnessPlayerLogic", world: "WitnessWorld", pre_picked_entities: Set[str]):
+        self.player_logic = player_logic
+        self.player_options = world.options
+        self.player_name = world.player_name
+        self.random = world.random
+
+        self.PRE_PICKED_HUNT_ENTITIES = pre_picked_entities.copy()
+        self.HUNT_ENTITIES = set()
+
+        self._add_plandoed_hunt_panels_to_pre_picked()
+
+        self.ALL_ELIGIBLE_ENTITIES, self.ELIGIBLE_ENTITIES_PER_AREA = self._get_eligible_panels()
+
+    def pick_panel_hunt_panels(self, total_amount: int) -> Set[str]:
+        """
+        The process of picking all hunt entities is:
+
+        1. Add pre-defined hunt entities
+        2. Pick random hunt entities to fill out the rest
+        3. Replace unfair entities with fair entities
+
+        Each of these is its own function.
+        """
+
+        self.HUNT_ENTITIES = self.PRE_PICKED_HUNT_ENTITIES.copy()
+
+        self._pick_all_hunt_entities(total_amount)
+        self._replace_unfair_hunt_entities_with_good_hunt_entities()
```

Comment:
This function should account for plando'ed Panel Hunt panels.

Reply:
It does.
Plandoed panels are added to `self.PRE_PICKED_HUNT_ENTITIES`
In this snapshot you can see: `self.HUNT_ENTITIES = self.PRE_PICKED_HUNT_ENTITIES.copy()`
And `_replace_unfair_hunt_entities_with_good_hunt_entities()` operates on `self.HUNT_ENTITES`.

Reply:
Oh, I just realized what you meant.
You want it to **not** replace plandoed stuff